excise/utils.py

```diff
@@ -0,0 +1,674 @@
+    )
+
+
+def get_order_request_data(order: "Order", config=AvataxConfiguration):
+    discount_total = get_total_order_discount(order)
+    discounted = True if discount_total.amount > 0 else False
+
+    lines = get_order_lines_data(order, discounted=discounted)
+    data = generate_request_data(
+        transaction_type=TRANSACTION_TYPE,
+        lines=lines,
+        invoice_number=f"{order.pk}",
+        discount=discount_total.amount,
+    )
+    return data
+
+
+def get_order_tax_data(
+    order: "Order", config: AvataxConfiguration, force_refresh=False
+) -> Dict[str, Any]:
+    data = get_order_request_data(order)
+
+    response = get_cached_response_or_fetch(
+        data, "order_%s" % order.token, config, force_refresh
+    )
+    if response.get("Status") != "Success":
+        transaction_errors = response.get("TransactionErrors")
+        customer_msg = ""
+        if isinstance(transaction_errors, list):
+            for error in transaction_errors:
+                error_message = error.get("ErrorMessage")
+                if error_message:
+                    customer_msg += error_message
+                error_code = response.get("ErrorCode", "")
+                logger.warning(
+                    "Unable to calculate taxes for order %s, error_code: %s, "
+                    "error_msg: %s",
+                    order.token,
+                    error_code,
+                    error_message,
+                )
+                if error_code == "-1003":
+                    raise ValidationError(error_message)
+        raise TaxError(customer_msg)
+    return response
+
+
+def _retrieve_meta_data_from_cache(token):
+    cached_data = cache.get(token)
+    return cached_data
+
+
+def metadata_requires_update(
+    metadata: str,
+    token_in_cache: str,
+    force_refresh: bool = False,
+):
+    """
+    Check if Checkout metadata needs to be reset.
+    The itemized taxes from ATE are stored in a cache.
+    If an object doesn't exist in cache or something has changed,
+    taxes need to be refetched.
+    """
+    if force_refresh:
+        return True
+
+    cached_metadata = _retrieve_meta_data_from_cache(token_in_cache)
+
+    if not cached_metadata:
+        return True
+
+    if cached_metadata != metadata:
+        return True
+
+    return False
+
+
+def process_checkout_metadata(
+    metadata: str,
+    checkout: "Checkout",
+    force_refresh: bool = False,
+    cache_time: int = CACHE_TIME,
+):
+    """
+    Check for Checkout metadata in cache.
+    Do nothing if metadata are the same. Set new metadata in other cases.
+    """
+    checkout_token = checkout.token
+    metadata_key = get_metadata_key("checkout_metadata_")
+    data_cache_key = f"{metadata_key}{checkout_token}"
+    tax_item = {get_metadata_key("itemized_taxes"): metadata}
+
+    if metadata_requires_update(tax_item, data_cache_key) or force_refresh:
+
+        checkout_obj = Checkout.objects.filter(token=checkout_token).first()
```

Comment:
Why do you retrieve Checkout from DB one more time? You already have the checkout object received as a function argument

Reply:
Fixed here: https://github.com/saleor/saleor-plugin-avatax-excise/pull/7/commits/4a1f66bffac73fecf01fa66ea702db2f2fdd3fe3